lib/connection/connect.js

```diff
@@ -0,0 +1,172 @@
+'use strict';
+const net = require('net');
+const tls = require('tls');
+const Connection = require('./connection');
+
+function connect(messageHandler, options, callback) {
+  if (options.family !== void 0) {
+    makeConnection(options.family, options, (err, socket) => {
+      if (err) {
+        return callback(err, socket); // in the error case, `socket` is the originating error event
+      }
+
+      callback(null, new Connection(messageHandler, socket, options));
+    });
+
+    return;
+  }
+
+  return makeConnection(6, options, (err, ipv6Socket) => {
+    if (err) {
+      // if (this.logger.isDebug()) {
+      //   this.logger.debug(
+      //     `connection ${this.id} for [${this.address}] errored out with [${JSON.stringify(err)}]`
+      //   );
+      // }
+
+      makeConnection(4, options, (err, ipv4Socket) => {
+        if (err) {
+          return callback(err, ipv4Socket); // in the error case, `ipv4Socket` is the originating error event
+        }
+
+        callback(null, new Connection(messageHandler, ipv4Socket, options));
+      });
+
+      return;
+    }
+
+    callback(null, new Connection(messageHandler, ipv6Socket, options));
+  });
+}
+
+const LEGAL_SSL_SOCKET_OPTIONS = [
+  'pfx',
+  'key',
+  'passphrase',
+  'cert',
+  'ca',
+  'ciphers',
+  'NPNProtocols',
+  'ALPNProtocols',
+  'servername',
+  'ecdhCurve',
+  'secureProtocol',
+  'secureContext',
+  'session',
+  'minDHSize'
+];
+
+function parseConnectOptions(family, options) {
+  const host = typeof options.host === 'string' ? options.host : 'localhost';
+  if (host.indexOf('/') !== -1) {
+    return { path: host };
+  }
+
+  const result = {
+    family,
+    host,
+    port: typeof options.port === 'number' ? options.port : 27017,
+    rejectUnauthorized: false
+  };
+
+  return result;
+}
+
+function parseSslOptions(family, options) {
+  const result = parseConnectOptions(family, options);
+
+  // Merge in valid SSL options
+  for (const name in options) {
+    if (options[name] != null && LEGAL_SSL_SOCKET_OPTIONS.indexOf(name) !== -1) {
+      result[name] = options[name];
+    }
+  }
+  // Set options for ssl
+  if (options.ca) result.ca = options.ca;
```

Comment:
I'd like to add a comment to investigate these assignments. I think they are probably superfluous now that we have the `for` loop above.